src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSnapshotChangeEventSource.java

```diff
@@ -498,9 +502,26 @@ protected SnapshotResult<YugabyteDBOffsetContext> doExecute(ChangeEventSourceCon
                   continue;
                 }
 
+                // Stores the READ records received in a single GetChanges call
+                long readRecordsReceived = 0;
+
                 CdcSdkCheckpoint explicitCdcSdkCheckpoint = null;
                 if (taskContext.shouldEnableExplicitCheckpointing()) {
-                  explicitCdcSdkCheckpoint = tabletToExplicitCheckpoint.get(part.getId());
+                  CdcSdkCheckpoint checkpoint = tabletToExplicitCheckpoint.get(part.getId());
+
+                  // If the checkpoint in tabletToExplicitCheckpoint map corresponds to the last record's checkpoint
+                  // of last snapshot batch, it means kafka has consumed the last record. So, we should
+                  // not poll on this tablet anymore, instead mark snapshot done for this tablet.
+                  if (checkpoint != null && isLastSnapshotRecordOfLastBatch(OpId.from(checkpoint))) {
+                    LOGGER.debug("Already received the last record's modified checkpoint in kafka's callback. " +
+                            "Discontinuing polling on tablet {}", part.getId());
+                    LOGGER.info("Adding {} to the list of snapshot completed tablets", part.getId());
+                    snapshotCompletedTablets.add(part.getId());
+                    markSnapshotDoneOnServer(part, previousOffset);
```

Comment:
only when markSnapshotDoneOnServer is complete we should do others. markSnapshotDone can throw error.

Reply:
Can you check the behaviour  if markSnapshotDoneOnServer fails?

Reply:
Checked the behavior - it can either fail to switch to streaming phase or succesfully switch to streaming depending on the number of tablets chosen for snapshot and the retry count. 

Based on our discussion, finally, changing it to this flow: First call `markSnapshotDoneOnServer()` and then only add the tablet to `snapshotCompletedTablets` set.